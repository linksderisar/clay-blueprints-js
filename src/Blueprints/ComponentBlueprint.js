import _ from 'lodash';
import Blueprint from './Blueprint';
import TextBlueprint from './TextBlueprint';

export const SLOT_PROPS = '$_slot_props';

export const SELF = '$_self';

export const LOOP = 'loop';

export const LOOP_VALUE = '$_loop_value';

const mapWithColon = (obj) => _.map(obj, (v, k) => ({[`:${k}`]: v}));

export default class ComponentBlueprint extends Blueprint {
    constructor(type) {
        super(type);

        this._ref = '';
        this._key = '';
        this._refInFor = false;
        this._scopedSlots = {};
        this._attributes = {};
        this._props = {};
        this._on = {};
        this._children = {};
        this._classes = [];
        this._style = {};
        this._text = null;
        this._loop = null;
        this._affect = '';
        this._bound = {};
        this._conditions = [];
    }

    addBound(bound) {
        this.bound = {
            ...this.bound,
            ...bound
        };

        return this;
    }

    set bindKey(key) {
        this.key = key;
        return this.addBound('key');
    }

    set bindRef(ref) {
        this.ref = ref;
        return this.addBound('ref');
    }

    get ref() {
        return this._ref;
    }

    set ref(value) {
        this._ref = value;
        return this;
    }

    get key() {
        return this._key;
    }

    set key(value) {
        this._key = value;
        return this;
    }

    get refInFor() {
        return this._refInFor;
    }

    set refInFor(value) {
        this._refInFor = value;
        return this;
    }

    get scopedSlots() {
        return this._scopedSlots;
    }

    set scopedSlots(cb) {
        const ret = cb((ref) => this.slotProp(ref));
        this._scopedSlots = Array.isArray(ret) ? ret : [ret];
        return this;
    }

    slotProp(ref) {
        return `${SLOT_PROPS}.${this.id}.${ref}`;
    }

    get attributes() {
        return this._attributes;
    }

    set attributes(value) {
        this._attributes = value;
        return this;
    }

    set bindAttributes(attributes) {
        this.attributes = mapWithColon(attributes);
        return this;
    }

    addBindAttribute(key, value) {
        return this.addBindAttributes({[key]: value});
    }

    addBindAttributes(attributes) {
        this.attributes = {
            ...this.attributes,
            ...mapWithColon(attributes)
        };

        return this;
    }

    addAttributes(attributes) {
        this.attributes = {
            ...this.attributes,
            ...attributes
        };

        return this;
    }

    get props() {
        return this._props;
    }

    set props(value) {
        this._props = value;
        return this;
    }

    set bindProps(props) {
        this.props = mapWithColon(props);
        return this;
    }

    addProps(props) {
        this.props = _.merge(this.props, props);
        return this;
    }

    addBindProps(props) {
        this.props = _.merge(this.props, mapWithColon(props));

        return this;
    }

    get on() {
        return this._on;
    }

    set on(value) {
        this._on = value;
        return this;
    }

    get children() {
        return this._children;
    }

    set children(value) {
        this._children = value;
        return this;
    }

    get classes() {
        return this._classes;
    }

    set classes(classes) {
        this._classes = {'class': classes};
        return this;
    }

    set bindClasses(ref) {
        this._classes = {':class': ref};
        return this;
    }

    get style() {
        return this._style;
    }

    set style(value) {
        this._style = value;
        return this;
    }

    get text() {
        return this._text;
    }

    set text(value) {
        this._text = new TextBlueprint(value);
        return this;
    }

    get loop() {
        return this._loop;
    }

    set loop(value) {
        this._loop = value;
        return this;
    }

    get affect() {
        return this._affect;
    }

    set affect(value) {
        this._affect = value;
        return this;
    }

    get bound() {
        return this._bound;
    }

    set bound(value) {
        this._bound = value;
        return this;
    }

    isBound(key) {
        return _.find(this.bound, (k) => k === key);
    }

    get conditions() {
        return this._conditions;
    }

    set conditions(value) {
        this._conditions = value;
        return this;
    }

    addCondition(condition) {
        this.conditions = [
            ...this.conditions,
            condition
        ];
    }

    toObject() {
        let obj = super.toObject();

        if (!obj.hasOwnProperty('attributes')) {
            obj['attributes'] = {};
        }

        if (this.key) {
            obj.attributes[this.isBound('key') ? ':key' : 'key'] = this.key;
        }

        if (this.ref) {
            obj.attributes[this.isBound('ref') ? ':ref' : 'ref'] = this.ref;
        }

        if (this.refInFor) {
            obj.attributes['refInFor'] = this.refInFor;
        }

        if (!_.isEmpty(this.style)) {
            obj.attributes['style'] = this.style;
        }

        if (!_.isEmpty(this.props)) {
            obj.attributes['props'] = this.props;
        }

        if (!_.isEmpty(this.on)) {
            obj.attributes['on'] = this.on;
        }

        if (!_.isEmpty(this.classes)) {
            obj.attributes = {
                ...obj.attributes,
                ...this.classes
            };

        }

        if (!_.isEmpty(this.attributes)) {
            obj.attributes['attrs'] = {
                ...obj.attributes,
                ...this.attributes
            };
        }

        if (this.loop) {
            obj = {
                ...obj,
                ...this.loop.toObject()
            };
        }

        if (!_.isEmpty(this.conditions)) {
            obj = {
                ...obj,
                ..._.map(this.conditions, (condition) => condition.toObject())
            };
        }

        if (this.text) {
            obj.children = this.text.toObject();
            return obj;
        }

        if (this.affect) {
            obj.affect = this.affect;
        }

        if (!_.isEmpty(this.scopedSlots)) {
            obj.scopedSlots = _.map(this.scopedSlots, s => s.toObject());
        }

        if (!_.isEmpty(this.children)) {
            obj.children = _.map(this.children, c => c.toObject());
        }

        if (_.isEmpty(obj.attributes)) {
            delete obj.attributes;
        }

        return obj;
    }
}
