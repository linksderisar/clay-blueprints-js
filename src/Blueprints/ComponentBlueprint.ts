import AbstractBlueprint from "./Abstracts/AbstractBlueprint";
import TextBlueprint from "./TextBlueprint";
import BlueprintException from "../Exceptions/BlueprintException";
import AbstractConditionBlueprint from "./Abstracts/AbstractConditionBlueprint";

export const SLOT_PROPS = '$_slot_props';

export const SELF = '$_self';

export const LOOP = 'loop';

export const LOOP_VALUE = '$_loop_value';

const mapWithColon = (obj: {}) => {
    const newObj = {};
    Object.keys(obj).forEach(k => newObj[":" + k] = obj[k]);
    return newObj;
};

export default class ComponentBlueprint extends AbstractBlueprint {
    protected ref = '';
    protected key = '';
    protected refInFor = false;
    protected scopedSlots: AbstractBlueprint[] = [];
    protected attributes = {};
    protected props = {};
    protected on = {};
    protected children: ComponentBlueprint[] = [];
    protected classes = {};
    protected style: string[] = [];
    protected affect = '';
    protected bound: string[] = [];
    protected loop: AbstractBlueprint|undefined;
    protected conditions: AbstractConditionBlueprint[] = [];
    protected text: TextBlueprint|undefined;

    // fixme add text, loop, and conditions

    public getBound(): string[] {
        return this.bound;
    }

    public setBound(bound: string[]): ComponentBlueprint {
        this.bound = bound;
        return this;
    }

    protected addBound(bound: string): ComponentBlueprint {
        this.bound = [...this.bound, bound];
        return this;
    }

    public getRef(): string {
        return this.ref;
    }

    public setRef(ref: string): ComponentBlueprint {
        this.ref = ref;
        return this;
    }

    public setBindRef(ref: string): ComponentBlueprint {
        this.ref = ref;
        this.addBound('ref');
        return this;
    }

    public getKey(): string {
        return this.key;
    }

    public setKey(key: string): ComponentBlueprint {
        this.key = key;
        return this;
    }

    public setBindKey(key: string): ComponentBlueprint {
        this.key = key;
        this.addBound('key');
        return this;
    }

    public getRefInFor(): boolean {
        return this.refInFor;
    }

    public setRefInFor(refInFor: boolean): ComponentBlueprint {
        this.refInFor = refInFor;
        return this;
    }

    public getScopedSlots(): AbstractBlueprint[] {
        return this.scopedSlots;
    }

    public setScopedSlots(scopedSlots: (string) => AbstractBlueprint|AbstractBlueprint[]): ComponentBlueprint {
        const bp: AbstractBlueprint|AbstractBlueprint[] = scopedSlots((ref) => this.slotProp(ref));

        if (bp instanceof AbstractBlueprint) {
            this.scopedSlots = [bp];
        } else {
            this.scopedSlots = bp;
        }

        return this;
    }

    public slotProp(ref: string): string {
        return `${SLOT_PROPS}.${this.getId()}.${ref}`;
    }

    public getAttributes(): {} {
        return this.attributes;
    }

    public setAttributes(attributes: {}): ComponentBlueprint {
        this.attributes = attributes;
        return this;
    }

    public setBindAttributes(attributes: {}): ComponentBlueprint {
        this.attributes = mapWithColon(attributes);
        return this;
    }

    public addBindAttribute(key: string, value: string) {
        return this.addBindAttributes({[key]: value});
    }

    public addBindAttributes(attributes: {}): ComponentBlueprint {
        this.attributes = {
            ...this.attributes,
            ...mapWithColon(attributes)
        };

        return this;
    }

    public addAttributes(attributes: {}): ComponentBlueprint {
        this.attributes = {
            ...this.attributes,
            ...attributes
        };

        return this;
    }

    public getProps(): {} {
        return this.props;
    }

    public setProps(props: {}): ComponentBlueprint {
        this.props = props;
        return this;
    }

    public setBindProps(props: {}): ComponentBlueprint {
        this.props = mapWithColon(props);
        return this;
    }

    public addProps(props: {}): ComponentBlueprint {
        // in Clay PHP props are replaced recursively, but this doesn't seem to make sense
        this.props = {
            ...this.props,
            ...props
        };
        return this;
    }

    public addBindProps(props: {}): ComponentBlueprint {
        this.props = {
            ...this.props,
            ...mapWithColon(props)
        };
        return this;
    }

    public addBindProp(key: string, value: any): ComponentBlueprint {
        return this.addBindProps({[key]: value});
    }

    public getOn(): {} {
        return this.on;
    }

    public setOn(on: {}): ComponentBlueprint {
        this.on = on;
        return this;
    }

    public addOn(on: {}): ComponentBlueprint {
        this.on = {
            ...this.on,
            ...on
        };
        return this;
    }

    public getChildren(): ComponentBlueprint[] {
        return this.children;
    }

    public setChildren(children: ComponentBlueprint[]): ComponentBlueprint {
        this.children = children;
        return this;
    }

    public getClasses(): {} {
        return this.classes;
    }

    public setClasses(classes: {}): ComponentBlueprint {
        this.classes = {'class': classes};
        return this;
    }

    public setBindClasses(ref: string): ComponentBlueprint {
        this.classes = {':class': ref};
        return this;
    }

    public getStyle(): string[] {
        return this.style;
    }

    public setStyle(style: string[]): ComponentBlueprint {
        this.style = style;
        return this;
    }

    public getLoop(): AbstractBlueprint|undefined {
        return this.loop;
    }

    public setLoop(loop: AbstractBlueprint): ComponentBlueprint {
        this.loop = loop;
        return this;
    }

    public getConditions(): AbstractConditionBlueprint[] {
        return this.conditions;
    }

    public setConditions(conditions: AbstractConditionBlueprint[]): ComponentBlueprint {
        this.conditions = conditions;
        return this;
    }

    public addCondition(condition: AbstractConditionBlueprint): ComponentBlueprint {
        this.conditions = [
            ...this.conditions,
            condition
        ];
        return this;
    }

    public getAffect(): string {
        return this.affect;
    }

    public setAffect(affect: string): ComponentBlueprint {
        this.affect = affect;
        return this;
    }

    public getText(): TextBlueprint|undefined {
        return this.text;
    }

    public setText(text: string): ComponentBlueprint {
        this.text = new TextBlueprint(text);
        return this;
    }

    public setBindText(text: string): ComponentBlueprint {
        this.text = new TextBlueprint(text);
        this.text.bind();
        return this;
    }

    public isBound(key) {
        return this.bound.find((k) => k === key);
    }

    static create<T = AbstractBlueprint>(this: { new(): T }, ...attributes) {
        if (attributes.length === 0 || typeof attributes[0] !== 'string') {
            throw new BlueprintException('First parameter for create must be the type');
        }

        return new this(attributes[0]);
    }

    constructor(type: string) {
        super();
        this.type = type;
    }

    toObject() {
        let obj = super.toObject();

        if (!obj.hasOwnProperty('attributes')) {
            obj['attributes'] = {};
        }

        if (this.key !== '') {
            obj['attributes'][this.isBound('key') ? ':key' : 'key'] = this.key;
        }

        if (this.ref !== '') {
            obj['attributes'][this.isBound('ref') ? ':ref' : 'ref'] = this.ref;
        }

        if (this.refInFor) {
            obj['attributes']['refInFor'] = this.refInFor;
        }

        if (this.style.length > 0) {
            obj['attributes']['style'] = this.style;
        }

        if (Object.keys(this.props).length > 0) {
            obj['attributes']['props'] = this.props;
        }

        if (Object.keys(this.on).length > 0) {
            obj['attributes']['on'] = this.on;
        }

        if (Object.keys(this.classes).length > 0) {
            obj['attributes'] = {
                ...obj['attributes'],
                ...this.classes
            };

        }

        if (Object.keys(this.attributes).length > 0) {
            obj['attributes']['attrs'] = {
                ...obj['attributes'],
                ...this.attributes
            };
        }

        if (typeof this.loop !== "undefined") {
            obj = {
                ...obj,
                ...this.loop.toObject()
            };
        }

        if (this.conditions.length > 0) {
            obj = {
                ...obj,
                ...this.conditions.map((condition) => condition.toObject())
            };
        }

        if (typeof this.text !== 'undefined') {
            obj['children'] = this.text.toObject();
            return obj;
        }

        if (this.affect !== '') {
            obj['affect'] = this.affect;
        }

        if (this.scopedSlots.length > 0) {
            obj['scopedSlots'] = this.scopedSlots.map(s => s.toObject());
        }

        if (this.children.length > 0) {
            obj['children'] = this.children.map(c => c.toObject());
        }

        if (Object.keys(obj['attributes']).length === 0) {
            delete obj['attributes'];
        }

        return obj;
    }
}
