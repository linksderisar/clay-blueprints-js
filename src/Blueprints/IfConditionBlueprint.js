import ConditionBlueprint from './ConditionBlueprint';

export default class IfConditionBlueprint extends ConditionBlueprint {
    constructor() {
        super('if');
    }
}
