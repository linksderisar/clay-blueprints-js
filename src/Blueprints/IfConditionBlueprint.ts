import AbstractConditionBlueprint from './Abstracts/AbstractConditionBlueprint';

export default class IfConditionBlueprint extends AbstractConditionBlueprint {

    static create(...attributes): IfConditionBlueprint {
        return new this();
    }

    type = 'if';
}
