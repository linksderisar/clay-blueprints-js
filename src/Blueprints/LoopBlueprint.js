import Blueprint from './Blueprint';

export default class LoopBlueprint extends Blueprint {
    constructor (iterable) {
        super();
        this._iterable = iterable;
    }

    get iterable() {
        return this._iterable;
    }

    set iterable(value) {
        this._iterable = value;
    }

    toObject() {
        return {
            'loop': this.iterable
        };
    }
}
