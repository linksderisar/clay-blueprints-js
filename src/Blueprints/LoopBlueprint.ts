import AbstractBlueprint from './Abstracts/AbstractBlueprint';

export default class LoopBlueprint extends AbstractBlueprint {

    iterable: string = '';

    public getIterable(): string {
        return this.iterable;
    }

    public setIterable(iterable: string): this {
        this.iterable = iterable;
        return this;
    }

    toObject(): any {
        return {
            loop: this.iterable,
        };
    }
}
