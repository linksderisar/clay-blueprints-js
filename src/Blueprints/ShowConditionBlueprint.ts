import AbstractConditionBlueprint from './Abstracts/AbstractConditionBlueprint';

export default class ShowConditionBlueprint extends AbstractConditionBlueprint {

    static create(...attributes): ShowConditionBlueprint {
        return new this();
    }

    type = 'show';
}
