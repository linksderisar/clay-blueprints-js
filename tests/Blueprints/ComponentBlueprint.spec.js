import ComponentBlueprint from '../../src/Blueprints/ComponentBlueprint';
import LoopBlueprint from '../../src/Blueprints/LoopBlueprint';
import IfConditionBlueprint from '../../src/Blueprints/IfConditionBlueprint';
import ShowConditionBlueprint from '../../src/Blueprints/ShowConditionBlueprint';

let blueprint = null;

beforeEach(() => blueprint = new ComponentBlueprint('test'));

function matchesSnapshot() {
    expect(blueprint.toObject()).toMatchSnapshot({
        id: expect.any(String)
    });
}

it('creates a default blueprint', () => {
    matchesSnapshot();
});

it('sets a ref', () => {
    blueprint.ref = '$ref';
    matchesSnapshot();
});

it('sets a key', () => {
    blueprint.key = '$key';
    matchesSnapshot();
});

it('sets refInFor', () => {
    blueprint.refInFor = true;
    matchesSnapshot();
});

it('sets classes', () => {
    blueprint.classes = ['css-class'];
    matchesSnapshot();
});

it('sets bound classes', () => {
    blueprint.bindClasses = 'class.bind';
    matchesSnapshot();
});

it('sets styles', () => {
    blueprint.style = {'color': 'black'};
    matchesSnapshot();
});

it('sets affect', () => {
    blueprint.affect = 'affect.ref';
    matchesSnapshot();
});

it('sets text', () => {
    blueprint.text = 'Dummy Text';
    expect(blueprint.text).toMatchSnapshot({
        id: expect.any(String)
    });
});

it('sets props', () => {
    blueprint.props = {'prop': 'propValue'};
    matchesSnapshot();

    blueprint.addProps({'addProp': 'propValue'});
    matchesSnapshot();
});

it('sets bound props', () => {
    blueprint.bindProps = {'prop': 'prop.ref'};
    matchesSnapshot();

    blueprint.addBindProps({'addProp': 'prop.ref'});
    matchesSnapshot();
});

it('sets attributes', () => {
    blueprint.attributes = {'id': 'ID'};
    matchesSnapshot();

    blueprint.addAttributes({'type': 'password'});
    matchesSnapshot();
});

it('sets scoped slots', () => {
    const slot = new ComponentBlueprint('slot');
    blueprint.scopedSlots = () => [slot];

    expect(blueprint.toObject().scopedSlots[0]).toMatchSnapshot({
        id: expect.any(String)
    });

    const slot2 = new ComponentBlueprint('slot2');
    blueprint.scopedSlots = () => [slot, slot2];

    expect(blueprint.toObject().scopedSlots[1]).toMatchSnapshot({
        id: expect.any(String)
    });

});

it('slotProp helper works', () => {
    const slot = new ComponentBlueprint('slot');
    blueprint.scopedSlots = (slotProps) => [
        slot.addProps({'prop': slotProps('ref.prop')})
    ];

    expect(blueprint.toObject().scopedSlots[0]).toMatchObject({
        id: slot.id,
        type: "slot",
        attributes: {
            props: {
                prop: `$_slot_props.${blueprint.id}.ref.prop`
            }
        }
    });
});

it('sets children', () => {
    blueprint.children = [
        new ComponentBlueprint('child'),
        new ComponentBlueprint('child2')
    ];
    expect(blueprint.toObject().children[0]).toMatchSnapshot({
        id: expect.any(String)
    });

    expect(blueprint.toObject().children[1]).toMatchSnapshot({
        id: expect.any(String)
    });
});

it('sets loop', () => {
    blueprint.loop = new LoopBlueprint('loop.ref');
    matchesSnapshot();
});

it('sets if', () => {
    const ifCondition = new IfConditionBlueprint();
    ifCondition.condition = 'a === b';
    blueprint.addCondition(ifCondition);
    matchesSnapshot();
});

it('sets show', () => {
    const showCondition = new ShowConditionBlueprint();
    showCondition.condition = 'a === b';
    blueprint.addCondition(showCondition);
    matchesSnapshot();
});

it('sets events', () => {
    blueprint.on = {'event': 'method()'};
    blueprint.addOn({'event2': 'method2()'});
    matchesSnapshot();
});
