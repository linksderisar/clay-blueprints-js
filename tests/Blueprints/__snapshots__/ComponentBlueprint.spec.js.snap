// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a default blueprint 1`] = `
Object {
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets a key 1`] = `
Object {
  "attributes": Object {
    "key": "$key",
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets a ref 1`] = `
Object {
  "attributes": Object {
    "ref": "$ref",
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets affect 1`] = `
Object {
  "affect": "affect.ref",
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets attributes 1`] = `
Object {
  "attributes": Object {
    "attrs": Object {
      "id": "ID",
    },
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets attributes 2`] = `
Object {
  "attributes": Object {
    "attrs": Object {
      "id": "ID",
      "type": "password",
    },
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets bound classes 1`] = `
Object {
  "attributes": Object {
    ":class": "class.bind",
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets bound props 1`] = `
Object {
  "attributes": Object {
    "props": Array [
      Object {
        ":prop": "prop.ref",
      },
    ],
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets bound props 2`] = `
Object {
  "attributes": Object {
    "props": Array [
      Object {
        ":addProp": "prop.ref",
        ":prop": "prop.ref",
      },
    ],
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets children 1`] = `
Object {
  "id": Any<String>,
  "type": "child",
}
`;

exports[`sets children 2`] = `
Object {
  "id": Any<String>,
  "type": "child2",
}
`;

exports[`sets classes 1`] = `
Object {
  "attributes": Object {
    "class": Array [
      "css-class",
    ],
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets if 1`] = `
Object {
  "0": Object {
    "if": "a === b",
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets loop 1`] = `
Object {
  "id": Any<String>,
  "loop": "loop.ref",
  "type": "test",
}
`;

exports[`sets props 1`] = `
Object {
  "attributes": Object {
    "props": Object {
      "prop": "propValue",
    },
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets props 2`] = `
Object {
  "attributes": Object {
    "props": Object {
      "addProp": "propValue",
      "prop": "propValue",
    },
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets refInFor 1`] = `
Object {
  "attributes": Object {
    "refInFor": true,
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets scoped slots 1`] = `
Object {
  "id": Any<String>,
  "type": "slot",
}
`;

exports[`sets scoped slots 2`] = `
Object {
  "id": Any<String>,
  "type": "slot2",
}
`;

exports[`sets show 1`] = `
Object {
  "0": Object {
    "show": "a === b",
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets styles 1`] = `
Object {
  "attributes": Object {
    "style": Object {
      "color": "black",
    },
  },
  "id": Any<String>,
  "type": "test",
}
`;

exports[`sets text 1`] = `
Object {
  "_bound": false,
  "_content": "Dummy Text",
  "_type": "$text",
  "id": Any<String>,
}
`;
